File: flink-hello-world/flink-1.18-hello-world/src/main/java/RandomCharSource.java
import org.apache.flink.streaming.api.functions.source.SourceFunction;

import java.util.Random;

public class RandomCharSource implements SourceFunction<String> {

    private volatile boolean isRunning = true;

    @Override
    public void run(SourceContext<String> ctx) throws Exception {
        Random random = new Random();

        while (isRunning) {
            // Generate a random character between 'a' and 'z'
            char randomChar = (char) (random.nextInt(26) + 'a');

            // Emit the character as a String
            ctx.collect(String.valueOf(randomChar));
            System.out.println("Generate random char: " + randomChar);

            // Wait for 1 second before generating the next character
            Thread.sleep(1000);
        }
    }

    @Override
    public void cancel() {
        isRunning = false;
    }
}

File: flink-connectors/flink-1.12-hello-world/src/main/scala/RandomCharSource.scala
import org.apache.flink.streaming.api.functions.source.SourceFunction
import scala.util.Random

class RandomCharSource extends SourceFunction[String] {
  @volatile private var isRunning = true

  override def run(ctx: SourceFunction.SourceContext[String]): Unit = {
    val random = new Random()
    while (isRunning) {
      // Generate a random character
      val randomChar = (random.nextInt(26) + 'a').toChar.toString
      // Emit the character
      ctx.collect(randomChar)
      println(s"Emitting character: $randomChar")
      // Wait for 1 second
      Thread.sleep(1000)
    }
  }

  override def cancel(): Unit = {
    isRunning = false
  }
}
